README
NAME
    Dpchrist::CGI - utility subroutines for CGI scripts

DESCRIPTION
    This documentation describes module revision $Revision: 1.43 $.

    This is alpha test level software and may change or disappear at any
    time.

  GLOBALS
   %CHECKBOX_ARGS
        %CHECKBOX_ARGS = ();

    Default argments used by gen_checkbox().

   $CHECKSUM_SALT
        $CHECKSUM_SALT = join ' ', __PACKAGE__, __FILE__, __LINE__;

    Default hashing salt used by gen_hidden_checksum(). Caller should set
    this value after use'ing this module.

   %PASSWORD_FIELD_ARGS
        %PASSWORD_FIELD_ARGS = (
            -size      => 70,
            -maxlength => 80,
        );

    Default argments used by gen_password().

   $RX_UNTAINT_CHECKBOX
        $RX_UNTAINT_CHECKBOX = qr/(on)/;

    Regular expression used for untainting paths.

   $RX_UNTAINT_PATH
        $RX_UNTAINT_PATH = qr/([^\x00]+)/;

    Regular expression used for untainting paths.

   $RX_UNTAINT_TEXTAREA
        $RX_UNTAINT_TEXTAREA = qr/([\PC\n\r]+)/;

    Regular expression used for untainting textareas.

   $RX_UNTAINT_TEXTFIELD
        $RX_UNTAINT_TEXTFIELD = qr/([\PC]+)/;

    Regular expression used for untainting textfields and password fields.

   %TD_ATTR
        %TD_ATTR = (
            -align  => 'LEFT', 
            -valign => 'TOP',
        );

    Default attributes used by gen_td().

   %TEXTAREA_ARGS
        TEXTAREA_ARGS = (    
            -maxlength => 32*1024,
            -columns   => 80,
            -rows      => 24,
            -wrap      => 'physical',
        );

    Default arguments used by gen_textarea().

   %TEXTFIELD_ARGS
        %TEXTFIELD_ARGS = (
            -size      => 70,
            -maxlength => 80,
        );

    Default arguments used by gen_textfield().

   %TH_ATTR
        %TH_ATTR = (
            -align  => 'LEFT', 
            -valign => 'TOP',
            -width  => '20%',
        );

    Default attributes used by gen_th().

  SUBROUTINES
   calc_checksum
        my $md5 = calc_checksum(LIST);

        # LIST are items to be fed into the checksum

    Walks list and expands array references. Passes through call to
    Digest::MD5::md5_hex() using $CHECKSUM_SALT and walked list.

    LIST items, and referenced array items, should be strings, otherwise the
    checksum seems to be different for each hit (?).

    Calls Carp::confess() on error.

   dump_cookies
        push @debug_html, pre(escapeHTML(dump_cookies()));

    Calls get_cookies_as_rhh() (see below), feeds the returned reference to
    Data::Dumper->Dump(), and returns the result.

   dump_params
        push @debug_html, pre(escapeHTML(dump_params()));

        push @debug_html, pre(escapeHTML(dump_params(OBJECT)));

        # OBJECT (optional) is a CGI object or CGI-derived object

    Calls get_params_as_rha() (see below), feeds the returned reference to
    Data::Dumper->Dump(), and returns the result.

   gen_checkbox
        push @html, gen_checkbox(ARGS);

        # ARGS are named arguments

    Merges named arguments ARGS with default arguments %CHECKBOX_ARGS (ARGS
    have priority), and passes through call with net arguments to
    CGI::checkbox().

   gen_hidden
        push @html, gen_hidden(-name => NAME, -value => VALUE);

        # NAME is a CGI parameter name

        # VALUE is the parameter value, or a reference to a list of values

    Returns an array of HTML elements:

    [0] A hidden control with name NAME and value VALUE.

    [1] A hidden control with name given by NAME with '_ck' suffix and value
    given by calling calc_checksum() with $CHECKSUM_SALT and incoming
    arguments.

    Calls Carp::confess() on error.

   gen_password_field
        push @html, gen_password_field(ARGS);

        # ARGS are named arguments

    Merges named arguments ARGS with default arguments %PASSWORD_FIELD_ARGS
    (ARGS have priority), and passes through call with net arguments to
    CGI::password_field().

   gen_td
        push @html, table( Tr([
            th('one') . gen_td(ARGS),
            th('two') . gen_td(ATTR, ARGS),
        ]));

        # ARGS are named arguments

        # ATTR (optional) is a reference to a hash of attributes

    Merges named attributes ATTR with default attributes %TD_ATTR (ATTR has
    priority) and passes through call with net attributes and arguments to
    CGI::td().

   gen_textarea
        push @html, gen_textarea(ARGS);

        # ARGS is a list of named arguments

    Merges name arguments ARGS with default arguments %TEXTAREA_ARGS (ARGS
    has priority) and passes through call with net arguments to
    CGI::textarea().

   gen_textfield
        push @html, gen_textfield(ARGS);

        # ARGS is a list of named arguments

    Merges named arguments ARGS with default arguments %TEXTAREA_ARGS (ARGS
    has priority) and passes through call with net attributes and arguments
    to CGI::textfield().

   gen_th
        push @html, table( Tr([
            gen_th(ARGS)       . td(1),
            gen_th(ATTR, ARGS) . td(2),
        ]));

        # ARGS is a list of arguments

        # ATTR (optional) is a reference to a hash of named attributes

    Merges named attributes ATTR with default attributes %TH_ATTR (ATTR has
    priority) and passes through call with net attributes and arguments to
    CGI::th().

   get_cookies_as_rhh
        my $cookies = get_cookies_as_rhh();

    Calls CGI::cookie() in list context for all CGI cookies and returns a
    reference to a hash-of-hashes data structure.

   get_params_as_rha
        my $params = get_params_as_rha();

        my $params = get_params_as_rha(OBJECT);

        # OBJECT (optional) is a CGI object or CGI-derived object

    Calls CGI::param() in list context for all CGI parameters and returns a
    hash-of-arrays data structure.

    Calls Carp::confess() on error.

   merge_args
        merge_args(ARRAYREF, ARGS);

        # ARRAYREF is a reference to an array

        # ARGS is a reference to a hash of named arguments

    Inserts or merges named arguments ARGS into array referenced by
    ARRAYREF. Typically used with CGI.pm widget generating functions, such
    as textarea(). Referenced array is modified in the process, and
    key/value pairs may be reordered. Returns void.

    Calls Carp::confess() on error.

   merge_attr
        merge_attr(ARRAYREF, ATTR);

        # ARRAYREF is a reference to an array

        # ATTR is a reference to a hash of named attributes

    Inserts or merges named attributes ATTR into array referenced by
    ARRAYREF. Attributes in first element of referenced array take
    precedence over attributes in ATTR. Typically used with CGI.pm tag
    generating functions, such as td(). First element of referenced array
    may be created or modified in the process. Returns void. Doesn't bother
    to merge if reference array is empty. Returns void.

    Calls Carp::confess() on error.

   nbsp
        push @html, nbsp();

        push @html, nbsp(EXPR);

        # EXPR (optional) is a whole number

    Returns one or more nonbreaking space HTML character entities.

    Calls Carp::confess() on error.

   untaint_checkbox
        my @untainted = untaint_checkbox(LIST);

        # LIST are strings to be untainted

    Passes through call to untaint_regex() using a RX suitable for checkboxs
    ('on').

   untaint_path
        my @untainted = untaint_path(LIST);

        # LIST are strings to be untainted

    Passes through call to untaint_regex() using a RX suitable for Unix
    paths (everying except NULL).

   untaint_regex
        my @untainted = untaint_regex(RX, LIST);

        # RX is a regular epxression

        # LIST are strings to be untainted

    Applies regular expression to each element in LIST. If LIST is empty,
    returns void. In list context, process LIST and return first captured
    substrings for each LIST item. In scalar context, process first LIST
    item and return first captured substring.

    Caller usually creates RX with the quote regular expression operator
    'qr()'.

    Returns undef for LIST elements that are references.

    Calls Carp::confess() on error.

    Calls Carp::cluck() if LIST contains undefined values or references.

   untaint_textarea
        my @untainted = untaint_textarea(LIST);

        # LIST are strings to be untainted

    Passes through call to untaint_regex() using a RX suitable for text
    areas (printable characters plus carriage return and line feed).

   untaint_textfield
        my @untainted = untaint_textfield(LIST);

        # LIST are strings to be untainted

    Passes through call to untaint_regex() using a RX suitable for
    textfields (printable characters).

   validate_checkbox
        push @errors, validate_checkbox(NAME);

        # NAME is a CGI parameter name

    Performs checkbox checks on CGI parameter with name NAME. Skips checks
    if CGI parameter does not exist (e.g. empty or fresh hit). Returns error
    strings if problems found.

    Calls Carp::confess() on error.

   validate_hidden
        push @errors, validate_hidden(NAME);

        # NAME is a CGI parameter name

    Performs hidden field checks on CGI parameter with name NAME. Skips
    checks if no CGI parameters exist (e.g. fresh hit). Returns error
    strings if problems found.

    Calls Carp::confess() on error.

   validate_parameter_present
        push @errors, validate_required_parameters(LIST);

        # LIST are CGI parameter names

    Checks that the CGI parameters with names in LIST are defined and have
    values. Skips checks if no CGI parameters exist (e.g. fresh hit).
    Returns list of error strings if problems found.

    Calls Carp::confess() on error.

   validate_password_field
        push @errors, validate_password_field(NAME);

        # NAME is a CGI parameter name

    Performs password field checks on CGI parameter with name NAME. Skips
    checks if CGI parameter does not exist (e.g. empty or fresh hit).
    Returns error strings if problems found.

    Calls Carp::confess() on error.

   validate_textarea
        push @error, validate_textarea(NAME);

        # NAME is a CGI parameter name

    Performs textarea checks on CGI parameter with name NAME. Skips checks
    if parameter does not exist (e.g. empty or fresh hit). Returns error
    strings if problems found.

    Calls Carp::confess() on error.

   validate_textfield
        push @errors, validate_textfield(NAME);

        # NAME is a CGI parameter name

    Performs textfield checks on CGI parameter with name NAME. Skips checks
    if parameter does not exist (e.g. empty or fresh hit). and returns error
    strings if problems found.

    Calls Carp::confess() on error.

  EXPORT
    None by default.

    All of the subroutines may be imported by using the ':all' tag:

        use Dpchrist::CGI           qw( :all );

    See 'perldoc Export' for everything in between.

INSTALLATION
    Old school:

        perl Makefile.PL
        make
        make test
        make install

    Minimal:

        cpan Dpchrist::CGI

    Complete:

        cpan Bundle::Dpchrist

  DEPENDENCIES
    See Makefile.PL in source distribution root directory.

SEE ALSO
        CGI.pm

AUTHOR
    David Paul Christensen dpchrist@holgerdanske.com

COPYRIGHT AND LICENSE
    Copyright (C) 2010 by David Paul Christensen

    This program is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by the
    Free Software Foundation; version 2.

    This program is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
    Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.



Changes
Revision history for Perl extension Dpchrist::CGI.

0.01  Fri Nov 24 09:55:54 2006
	- original version; created by h2xs 1.23 with options
		-XAn Dpchrist::CGI



Makefile.PL
#######################################################################
# $Id: Makefile.PL,v 1.9 2010-11-18 22:52:58 dpchrist Exp $
#
# Copyright 2010 by David Paul Christensen dpchrist@holgerdanske.com
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; version 2.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307,
# USA.
#######################################################################

use 5.010000;
use ExtUtils::MakeMaker;

use Dpchrist::Module::MakefilePL (
    -mcpani		=> $ENV{CPAN_AUTHORID},
    -pod2html	 	=> [qw(
	lib/Dpchrist/CGI.pm
    )],
    -release		=> $ENV{RELEASE_ROOT},
);

WriteMakefile(
    NAME          => 'Dpchrist::CGI',
    VERSION_FROM  => 'lib/Dpchrist/CGI.pm',
    ABSTRACT_FROM => 'lib/Dpchrist/CGI.pm',
    PREREQ_PM     => {
	'Dpchrist::Debug'	=> 1.039,
	'Dpchrist::Is'		=> 1.017,
	'Dpchrist::LangUtil'	=> 1.005,
	'Dpchrist::Module'	=> 1.018,
	'Capture::Tiny'		=> 0.06,
	'feature'		=> 0,
    },
    AUTHOR        => 'David Paul Christensen ' .
		     '<dpchrist@holgerdanske.com>',
    clean		=> {
	FILES => q( *~tmp */*~tmp Dpchrist* debug.log ),
    },
);

#######################################################################


lib/Dpchrist/CGI/Widget.pm
#######################################################################
# $Id: Widget.pm,v 1.4 2010-11-16 05:28:51 dpchrist Exp $
#######################################################################
# package:
#----------------------------------------------------------------------

package Dpchrist::CGI::Widget;

use 5.010;
use strict;
use warnings;

use constant DEBUG		=> 0;

our $VERSION = sprintf "%d.%03d", q$Revision: 1.4 $ =~ /(\d+)/g;

#######################################################################
# uses:
#----------------------------------------------------------------------

use Carp;
use CGI				qw( :standard );
use Data::Dumper;
use Dpchrist::Debug		qw( :all );
use Dpchrist::Is		qw( :all );

#######################################################################

=head1 NAME

Dpchrist::CGI::Widget - base class for CGI widgets

=head2 OBJECT METHODS

=cut

#----------------------------------------------------------------------

=head3 name

    OBJECT->name

Returns -name attribute of object.
Cannot be used to set attribute.

Calls Carp::confess() on error.

=cut

sub name
{
    ddump('call', [\@_], [qw(*_)]) if DEBUG;

    my $self = shift;
    confess join(' ',
	'ERROR: object method called incorrectly',
	Data::Dumper->Dump([\@_], [qw(*_)]),
    ) unless isa_object($self, __PACKAGE__);
    confess join(' ',
	"ERROR: attribute 'name' is read only",
	Data::Dumper->Dump([\@_], [qw(*_)]),
    ) if @_;

    my $name = $self->{-name};

    ddump('return', [$name], [qw(name)]) if DEBUG;
    return $name;
}

#----------------------------------------------------------------------

=head3 value

    OBJECT->value
    OBJECT->value EXPR

If EXPR provided,
sets CGI parameter per OBJECT -name to -value.
Returns (existing or updated) CGI parameter.
Semantics match CGI::param() as much as possible.
See 'perldoc CGI' for details.

Calls Carp::confess() on error.

=cut

sub value
{
    ddump('call', [\@_], [qw(*_)]) if DEBUG;

    my $self = shift;
    confess join(' ',
	'ERROR: object method called incorrectly',
	Data::Dumper->Dump([\@_], [qw(*_)]),
    ) unless isa_object($self, __PACKAGE__);

    confess join(' ',
	'ERROR: accepts zero or one arguments',
	Data::Dumper->Dump([\@_], [qw(*_)]),
    ) if 1 < @_;

    my $name = $self->name;

    CGI::param(
	-name => $name,
	-value => $_[0],
    ) if @_;

    if (wantarray) {
	my @value = CGI::param($name);
	ddump('return', [\@value], [qw(*value)]) if DEBUG;
	return @value;
    }
    else
    {
	my $value = CGI::param($name);
	ddump('return', [$value], [qw(value)]) if DEBUG;
	return $value;
    }
}

#######################################################################
# end of code:
#----------------------------------------------------------------------

1;

__END__

#######################################################################

=head1 INSTALLATION

Installed as part of Dpchrist::CGI.


=head1 SEE ALSO

    CGI
    Dpchrist::CGI
    Dpchrist::CGI::Widget::Checkbox


=head1 AUTHOR

David Paul Christensen dpchrist@holgerdanske.com


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 by David Paul Christensen

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 2.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307,
USA.

=cut

#######################################################################


lib/Dpchrist/CGI/Widget/Checkbox.pm
#######################################################################
# $Id: Checkbox.pm,v 1.5 2010-11-16 05:28:51 dpchrist Exp $
#######################################################################
# package:
#----------------------------------------------------------------------

package Dpchrist::CGI::Widget::Checkbox;

use 5.010;
use strict;
use warnings;

use constant DEBUG		=> 0;

use constant ATTR_NAME_DEFAULT	=> 'checkbox';
use constant CGI_CHECKBOX_ARGS_ALLOWED => {
    -checked	=> 1,
    -force	=> 1,
    -label	=> 1,
    -name	=> 1,
    -on		=> 1,
    -onClick	=> 1,
    -override	=> 1,
    -selected	=> 1,
    -value	=> 1,
};
use constant GENHTML_WIDGET_ARGS_ALLOWED => {
    -checked	=> 1,
    -force	=> 1,
    -label	=> 1,
    -name	=> 0,
    -on		=> 1,
    -onClick	=> 1,
    -override	=> 1,
    -selected	=> 1,
    -value	=> 1,
};
use constant NEW_ARGS_ALLOWED => {
    -checked	=> 1,
    -force	=> 1,
    -label	=> 1,
    -name	=> 1,
    -on		=> 1,
    -onClick	=> 1,
    -override	=> 1,
    -selected	=> 1,
    -value	=> 1,
};

#######################################################################
# uses:
#----------------------------------------------------------------------

use Carp;
use CGI				qw( :standard );
use Data::Dumper;
use Dpchrist::CGI::Widget;
use Dpchrist::Debug		qw( :all );
use Dpchrist::Is		qw( :all );

#######################################################################
# globals:
#----------------------------------------------------------------------

our %Checkbox = (
    -object_serial	=> 0,
);

our @ISA = qw(Dpchrist::CGI::Widget);

our $VERSION = sprintf "%d.%03d", q$Revision: 1.5 $ =~ /(\d+)/g;

#######################################################################

=head1 NAME

Dpchrist::CGI::Widget::Checkbox - derived class for CGI checkboxes

=head2 CONSTRUCTOR

=head3 new

    CLASS->new [KEY=>VALUE]...

Creates a checkbox object with specified attributes.
The following attributes are recognized:

    -checked
    -force
    -label
    -name
    -on	
    -onClick
    -override
    -selected
    -value

If -name is omitted,
constructor sets attribute to "textboxN";
where N is an automatically incrementing serial number
starting at 0.

If -value is specified,
calls CGI::param() and sets the -name parameter to the value specified.

Calls Carp::confess() on error.

=cut

sub new
{
    ddump('call', [\@_], [qw(*_)]) if DEBUG;

    my $class = shift;
    confess join (' ',
	'ERROR: class method called incorrectly',
	Data::Dumper->Dump([\@_], [qw(*_)]),
    ) unless isa_class($class, __PACKAGE__);

    my %args = @_;
    foreach (keys %args) {
	my @e = grep {!NEW_ARGS_ALLOWED->{$_}} keys %args;
	confess join(' ',
    	    'ERROR: argument(s) unknown or not allowed',
	    Data::Dumper->Dump([\@e], [qw(e)]),
	) if @e;
    }

    my $self = {};

    $self->{-name} = exists($args{-name})
	? $args{-name}
	: ATTR_NAME_DEFAULT . $Checkbox{-object_serial}++;

    CGI::param(
	-name  => $self->{-name},
	-value => $args{-value},
    ) if exists $args{-value};

    foreach (qw(
	-checked -selected -on -label -onClick -override -force )) {
	$self->{$_} = $args{$_} if exists $args{$_};
    }

    bless ($self, $class);

    ddump('return', [$self], [qw(self)]) if DEBUG;
    return $self;
}

#######################################################################

=head2 OBJECT METHODS

=head3 genhtml_widget

    OBJECT->genhtml_widget [KEY=>VALUE]...

Returns HTML fragment for displaying a checkbox widget in a web page.
Provided attributes, if any,
temporarily override object attributes of the same name.
The following attributes are recognized:

    -checked
    -force
    -label
    -on	
    -onClick
    -override
    -selected
    -value

Calls Carp::confess() on error.

=cut

# object methods:
#----------------------------------------------------------------------

sub genhtml_widget
{
    ddump('call', [\@_], [qw(*_)]) if DEBUG;

    my $self = shift;
    confess join(' ',
	'ERROR: object method called incorrectly',
	Data::Dumper->Dump([\@_], [qw(*_)]),
    ) unless isa_object($self, __PACKAGE__);

    my %args = @_;
    foreach (keys %args) {
	my @e = grep {!GENHTML_WIDGET_ARGS_ALLOWED->{$_}} keys %args;
	confess join(' ',
    	    'ERROR: argument(s) unknown or not allowed',
	    Data::Dumper->Dump([\@e], [qw(e)]),
	) if @e;
    }

    my %h = (%$self, %args);
    ddump([\%h], [qw(h)]) if DEBUG;

    my @html = CGI::checkbox(%h);

    ddump('return', [\@html], [qw(*html)]) if DEBUG;
    return @html;
}

#######################################################################
# end of code:
#----------------------------------------------------------------------

1;

__END__

#######################################################################

=head1 INSTALLATION

Installed as part of Dpchrist::CGI.


=head1 SEE ALSO

    CGI
    Dpchrist::CGI
    Dpchrist::CGI::Widget


=head1 AUTHOR

David Paul Christensen dpchrist@holgerdanske.com


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 by David Paul Christensen

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 2.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307,
USA.

=cut

#######################################################################


lib/Dpchrist/CGI.pm
#######################################################################
# $Id: CGI.pm,v 1.25 2010-11-19 00:58:42 dpchrist Exp $
#######################################################################
# package:
#----------------------------------------------------------------------

package Dpchrist::CGI;

use 5.010;
use strict;
use warnings;

use constant DEBUG		=> 0;

use constant RX_UNTAINT_PATH	=> qr/([^\x00]+)/;
use constant RX_UNTAINT_TEXTAREA => qr/([\PC\n\r]+)/;
use constant RX_UNTAINT_TEXTFIELD => qr/([\PC]+)/;

require Exporter;

our @ISA = qw(Exporter);

our %EXPORT_TAGS = ( 'all' => [ qw(
	dump_cookies
	dump_params
	get_params_as_rha
	get_cookies_as_rhh
	nbsp
	untaint_path
	untaint_regex
	untaint_textarea
	untaint_textfield
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw( );

our $VERSION = sprintf "%d.%03d", q$Revision: 1.25 $ =~ /(\d+)/g;

#######################################################################
# uses:
#----------------------------------------------------------------------

use Carp			qw( cluck confess );
use CGI				qw( :standard );
use Data::Dumper;
use Dpchrist::Debug		qw( :all );
use Dpchrist::Is		qw( :all );

#######################################################################

=head1 NAME

Dpchrist::CGI - utility subroutines for CGI scripts

=head1 SYNOPSIS

See synopsis.cgi in the source distribution cgi-bin/ directory:

    #! /usr/bin/perl -T
    use 5.010;
    use strict;
    use warnings;

    use CGI                         qw( :standard );
    use Dpchrist::CGI               qw( :all );

    my $pinfo   = untaint_path path_info;
    my $name    = untaint_textfield param('name');
    my $comment = untaint_textarea param('comment');
    my $special = untaint_regex(qr/([a-z ]+)/, param('special'));
    my $cookie  = cookie(-name => 'mycookie', -value => 'chololate chip');

    print(
	header(-cookie => $cookie),
	start_html,
	start_form,
	'name:',     nbsp(4), textfield(-name => 'name'   ), br,
	'comment: ',          textarea( -name => 'comment'), br,
	'special: ', nbsp(2), textfield(-name => 'special'), br,
	submit, br,
	a({
	    -href => script_name . '/path/info?' . join('&',
		'name=<strong>attempted%20sneaky%20tags</strong>',
		'comment=One%20bright%20day%0D%0A' .
			'In%20the%20middle%20of%20the%20night,',
		'special=Regex%20matches%20lowercase%20and%20whitespacE',
		),
	    },
	    'test link'
	), br,
	end_form, hr,
	'path info: ',       escapeHTML($pinfo),    br,
	'name:',    nbsp(4), escapeHTML($name),     br,
	'comment:',      pre(escapeHTML($comment)), br,
	'special:', nbsp(2), escapeHTML($special),  br,
	pre(
	    escapeHTML(dump_cookies),
	    escapeHTML(dump_params),
	), br,
	end_html,
    );


=head2 SUBROUTINES

=cut

#----------------------------------------------------------------------

=head3 dump_cookies

    dump_cookies

Calls get_cookies_as_rhh() (see below),
feeds the returned reference to Data::Dumper->Dump(),
and returns the result.

=cut

sub dump_cookies
{
    my $cookies = get_cookies_as_rhh();

    return Data::Dumper->Dump([$cookies], [qw(cookies)]);
}

#----------------------------------------------------------------------

=head3 dump_params

    dump_params
    dump_params OBJECT

Calls get_params_as_rha() (see below),
feeds the returned reference to Data::Dumper->Dump(),
and returns the result.

If OBJECT is provided,
it must be a CGI object or derived from CGI.

=cut

sub dump_params
{
    my $params = get_params_as_rha(@_);

    return Data::Dumper->Dump([$params], [qw(params)]);
}

#----------------------------------------------------------------------

=head3 get_cookies_as_rhh

    get_cookies_as_rhh

Calls CGI::cookie($i) in list context for all CGI cookies,
populating a hash-of-hashes data structure along the way,
and returns a reference to the data structure.

=cut

sub get_cookies_as_rhh
{
    my $cookies;

    foreach (cookie()) {
	$cookies->{$_} = { cookie($_) };
    }

    return $cookies;
}

#----------------------------------------------------------------------

=head3 get_params_as_rha

    get_params_as_rha
    get_params_as_rha OBJECT

Calls CGI::param($i) in list context for all CGI parameters,
populating a hash-of-arrays data structure along the way,
and returns a reference to the data structure.

If OBJECT is provided,
it must be a CGI object or derived from CGI.

=cut

sub get_params_as_rha
{
    ### $CGI::Q may be undefined if no CGI calls made yet

    my $q = shift;

    my @params = $q ? $q->param() : param();

    $q = $CGI::Q unless $q;

    confess join(' ', 'ERROR: bad CGI object',
	Data::Dumper->Dump([\@_, \@params, $q], [qw(*_ params q)]),
    ) unless isa_object($q, 'CGI');

    my $rha = {};

    foreach (@params) {
	$rha->{$_} = [ $q->param($_) ];
    }

    return $rha;
}

#----------------------------------------------------------------------

=head3 nbsp

    nbsp EXPR

Returns EXPR nonbreaking space HTML character entities.

=cut

sub nbsp
{
    my $n = shift || 1;

    my $s = '&nbsp; ';

    return $s x $n;
}

#----------------------------------------------------------------------

=head3 untaint_path

    untaint_path LIST

Passes through call to untaint_regex()
using a RX suitable for Unix paths
(everying except NULL).

=cut

sub untaint_path
{
    ddump('enter', [\@_], [qw(*_)]) if DEBUG;

    dprint('passing through call to untaint_regex()') if DEBUG;
    return untaint_regex(RX_UNTAINT_PATH, @_);
}

#----------------------------------------------------------------------

=head3 untaint_regex

    untaint_regex RX,LIST

Apply regular expression to each element in LIST.
If LIST is empty, return void.
In list context, process LIST
and return first captured substrings for each LIST item.
In scalar context, process first LIST item
and return first captured substring.

Caller usually creates RX with the quote regex operator 'qr()'.

Returns undef for LIST elements that are references.

Calls Carp::confess() on error.

=cut

sub untaint_regex
{
    ddump('enter', [\@_], [qw(*_)]) if DEBUG;

    confess join(' ',
	'ERROR: first argument must be a regular expression',
	Data::Dumper->Dump([\@_], [qw(*_)]),
    ) unless @_ && ref($_[0]) eq 'Regexp';

    my $rx = shift;

    if (@_) {
	my @r;
	foreach (@_) {
	    cluck join(' ',
		'attempt to apply regular expression',
		'to undefined value',
	    ) unless defined($_);
	    cluck 'attempt to apply regular expression to reference'
		if ref($_);
	    if (defined($_) && !ref($_)) {
		$_ =~ $rx;
		push @r, $1;
	    }
	    else {
		push @r, undef;
	    }
	    last unless wantarray;
	}
	ddump('return',
	    wantarray ? ([\@r],   ['*r']  )
		      : ([$r[0]], ['r[0]'])
	) if DEBUG;
	return (wantarray) ? @r : $r[0];
    }
    dprint 'returning void' if DEBUG;
    return;
}

#----------------------------------------------------------------------

=head3 untaint_textarea

    untaint_textarea LIST

Passes through call to untaint_regex()
using a RX suitable for text areas
(printable characters plus carriage return and line feed).

=cut

sub untaint_textarea
{
    ddump('enter', [\@_], [qw(*_)]) if DEBUG;

    dprint('passing through call to untaint_regex()') if DEBUG;
    return untaint_regex(RX_UNTAINT_TEXTAREA, @_);
}

#----------------------------------------------------------------------

=head3 untaint_textfield

    untaint_textfield LIST

Passes through call to untaint_regex()
using a RX suitable for text areas
(printable characters).

=cut

sub untaint_textfield
{
    ddump('enter', [\@_], [qw(*_)]) if DEBUG;

    dprint('passing through call to untaint_regex()') if DEBUG;
    return untaint_regex(RX_UNTAINT_TEXTFIELD, @_);
}

#######################################################################
# end of module:
#----------------------------------------------------------------------

1;

__END__

#######################################################################

=head2 EXPORT

None by default.

All of the subroutines may be imported by using the ':all' tag:

    use Dpchrist::CGI		qw( :all ); 

See 'perldoc Export' for everything in between.


=head1 INSTALLATION

Minimal:

    cpan Dpchrist::CGI

Complete:

    cpan Bundle::Dpchrist


=head1 SEE ALSO

    CGI.pm


=head1 AUTHOR

David Paul Christensen dpchrist@holgerdanske.com


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 by David Paul Christensen

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 2.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307,
USA.

=cut

#######################################################################
